# 1. Preprocessing

This section contains the following: 

    1.1. Overview of the dataset
    1.2. Addressing missing values
    1.3. Filtering required columns
    1.4. Adding required columns
    1.5. Offsetting
    1.6. Downloading preprocessing dataset as a CSV file
    1.7. Insights

We start by importing the required packages

In [2]:
import pandas as pd
import numpy as np

Next, the downloaded dataset *Analysis* is read into a Pandas DataFrame *preprocessing_dataset*. 

In [9]:
preprocessing_dataset = pd.read_csv("C:\\Users\\Sharmila\\Documents\\Earthquake Project\\Earthquake dataset.csv")

## <font color=black>1.1. Overview of the Dataset</font>

The columns in the dataset, their types, index values, first and last five rows is displayed. The describe() function is used to provide a quick statistical overview of the data.

In [10]:
# The names of the columns

columns_dataset = preprocessing_dataset.columns
print(columns_dataset)
print(len(columns_dataset))

Index(['Year', 'Mo', 'Dy', 'Hr', 'Mn', 'Sec', 'Tsu', 'Vol', 'Addl', 'Name',
       'Latitude', 'Longitude', 'Focal', 'Mag', 'MMI Int', 'Num', 'De',
       'Num.1', 'De.1', '$Mill', 'De.2', 'Num.2', 'De.3', 'Num.3', 'De.4',
       'Unnamed: 25'],
      dtype='object')
26


In [11]:
# The data type of each column

print(preprocessing_dataset.dtypes)

Year             int64
Mo             float64
Dy             float64
Hr             float64
Mn             float64
Sec            float64
Tsu             object
Vol             object
Addl            object
Name            object
Latitude       float64
Longitude      float64
Focal          float64
Mag            float64
MMI Int        float64
Num            float64
De             float64
Num.1          float64
De.1           float64
$Mill          float64
De.2           float64
Num.2          float64
De.3           float64
Num.3          float64
De.4           float64
Unnamed: 25    float64
dtype: object


In [12]:
# Next, the indexes associated with the dataframe are checked

preprocessing_dataset.index.values

array([   0,    1,    2, ..., 6141, 6142, 6143], dtype=int64)

The top five and bottom five rows are displayed below. It is seen that each observation forms a row, is defined by an index, and attributes of those observations are displayed in the columns of the DataFrame. 

In [13]:
# First five rows

print(preprocessing_dataset.head())

   Year  Mo  Dy  Hr  Mn  Sec  Tsu  Vol Addl                              Name  \
0 -2150 NaN NaN NaN NaN  NaN  NaN  NaN    *      JORDAN: BAB-A-DARAA,AL-KARAK   
1 -2000 NaN NaN NaN NaN  NaN  Tsu  NaN    *                     SYRIA: UGARIT   
2 -2000 NaN NaN NaN NaN  NaN  NaN  NaN    *                   TURKMENISTAN: W   
3 -1610 NaN NaN NaN NaN  NaN  Tsu  Vol    *  GREECE: THERA ISLAND (SANTORINI)   
4 -1566 NaN NaN NaN NaN  NaN  NaN  NaN    *           ISRAEL: ARIHA (JERICHO)   

   ...   De  Num.1  De.1  $Mill  De.2  Num.2  De.3  Num.3  De.4  Unnamed: 25  
0  ...  NaN    NaN   NaN    NaN   3.0    NaN   NaN    NaN   NaN          NaN  
1  ...  3.0    NaN   NaN    NaN   NaN    NaN   NaN    NaN   NaN          NaN  
2  ...  1.0    NaN   NaN    NaN   1.0    NaN   1.0    NaN   NaN          NaN  
3  ...  NaN    NaN   NaN    NaN   NaN    NaN   NaN    NaN   NaN          NaN  
4  ...  NaN    NaN   NaN    NaN   3.0    NaN   NaN    NaN   NaN          NaN  

[5 rows x 26 columns]


In [14]:
# Last five rows

print(preprocessing_dataset.tail())

      Year   Mo    Dy    Hr    Mn   Sec  Tsu  Vol Addl  \
6139  2019  9.0  24.0  11.0   1.0  55.0  NaN  NaN    *   
6140  2019  9.0  25.0  23.0  46.0  44.0  NaN  NaN    *   
6141  2019  9.0  26.0  10.0  59.0  26.0  NaN  NaN    *   
6142  2019  9.0  26.0  16.0  36.0  18.0  NaN  NaN    *   
6143  2019  9.0  29.0  15.0  57.0  53.0  NaN  NaN    *   

                           Name  ...   De  Num.1  De.1  $Mill  De.2  Num.2  \
6139  PAKISTAN: MIRPUR DISTRICT  ...  1.0  746.0   3.0    NaN   2.0  135.0   
6140   INDONESIA: MALUKU: AMBON  ...  1.0  179.0   3.0    NaN   3.0    NaN   
6141           TURKEY: ISTANBUL  ...  1.0   34.0   1.0    NaN   2.0    NaN   
6142       CHILE: SOUTH CENTRAL  ...  1.0    NaN   NaN    NaN   NaN    NaN   
6143          CHILE: CONCEPCION  ...  1.0    NaN   NaN    NaN   NaN    NaN   

      De.3   Num.3  De.4  Unnamed: 25  
6139   3.0   319.0   3.0          NaN  
6140   1.0  2675.0   2.0          NaN  
6141   NaN   473.0   3.0          NaN  
6142   NaN     NaN   N

In [15]:
# The basic statistical overview of the dataset is provided using Pandas describe()

print(preprocessing_dataset.describe(include = 'all'))

               Year           Mo           Dy           Hr          Mn  \
count   6144.000000  5739.000000  5587.000000  4113.000000  3908.00000   
unique          NaN          NaN          NaN          NaN         NaN   
top             NaN          NaN          NaN          NaN         NaN   
freq            NaN          NaN          NaN          NaN         NaN   
mean    1804.691243     6.505663    15.721675    11.304644    28.86131   
std      376.404999     3.449445     8.749653     7.032811    17.14624   
min    -2150.000000     1.000000     1.000000     0.000000     0.00000   
25%     1820.000000     4.000000     8.000000     5.000000    14.75000   
50%     1928.000000     7.000000    16.000000    11.000000    30.00000   
75%     1988.000000     9.000000    23.000000    17.000000    44.00000   
max     2019.000000    12.000000    31.000000    23.000000    59.00000   

                Sec   Tsu  Vol  Addl                    Name  ...  \
count   2791.000000  1853   64  6144      

## 1.2. Addressing Missing Values

The missing values in *Name, Latitude, Longitude*, and *Magnitude* are explored in detail. A question mark(?) is used to replace the missing values in other variables.  

In [16]:
# The null values in the dataset is viewed: 

print(preprocessing_dataset.isna().sum())

Year              0
Mo              405
Dy              557
Hr             2031
Mn             2236
Sec            3353
Tsu            4291
Vol            6080
Addl              0
Name              1
Latitude         54
Longitude        50
Focal          2951
Mag            1783
MMI Int        3379
Num            4104
De             3634
Num.1          4932
De.1           4747
$Mill          5648
De.2           1757
Num.2          5375
De.3           4475
Num.3          5675
De.4           5250
Unnamed: 25    6083
dtype: int64


### 1.2.1. Name

There exists one missing entry in *Name*. The row containing missing value is retrieved. Based on the latitude and longitude of mentioned in that row, the entire dataset is searched to find a similar entry.

In [17]:
# Accessing index of the missing entry in the Name column

np.where(pd.isnull(preprocessing_dataset['Name']))

(array([2028], dtype=int64),)

In [18]:
# Accessing the entire row of index 202

preprocessing_dataset.iloc[2028]

Year           1868
Mo               10
Dy               18
Hr               12
Mn               35
Sec             NaN
Tsu             Tsu
Vol             NaN
Addl              *
Name            NaN
Latitude      -40.2
Longitude       173
Focal            12
Mag             7.6
MMI Int         NaN
Num             NaN
De              NaN
Num.1           NaN
De.1            NaN
$Mill           NaN
De.2            NaN
Num.2           NaN
De.3            NaN
Num.3           NaN
De.4            NaN
Unnamed: 25     NaN
Name: 2028, dtype: object

In [19]:
# Finding similar rows with latitude -40.2 and longitude 173

same_latitude = preprocessing_dataset[preprocessing_dataset['Latitude'] == -40.2] 
print(same_latitude)

      Year    Mo    Dy    Hr    Mn  Sec  Tsu  Vol Addl Name  ...  De  Num.1  \
2028  1868  10.0  18.0  12.0  35.0  NaN  Tsu  NaN    *  NaN  ... NaN    NaN   

      De.1  $Mill  De.2  Num.2  De.3  Num.3  De.4  Unnamed: 25  
2028   NaN    NaN   NaN    NaN   NaN    NaN   NaN          NaN  

[1 rows x 26 columns]


The Name column is NaN as well. Google Maps is used to find the exact location to be Cook Strait in New Zealand.

In [20]:
# Assigning 'NEW ZEALAND: COOK STRAIT' to the missing value in 'Name'

preprocessing_dataset["Name"].fillna("NEW ZEALAND: COOK STRAIT", inplace = True)  

In [21]:
# Verifying if the null value still exists

print(preprocessing_dataset.Name.isna().sum())

0


### 1.2.2. Latitude and Longitude

In [22]:
# Calculating the total number of missing latitudes

print(preprocessing_dataset.Latitude.isna().sum())

54


In [23]:
# Calculating the total number of missing longitudes

print(preprocessing_dataset.Longitude.isna().sum())

50


The reasons behind the missing values is not yet understood. NOAA mentions that the unknown latitudes and longitudes are represented by '0.0'. Since there are 54 unknown latitudes and 50 unknown longitudes still present, all of them are replaced by '0.0'. 

In [24]:
# Replacing missing latitude values with 0.0

preprocessing_dataset["Latitude"].fillna("0.0", inplace = True)  

In [25]:
# Replacing missing longitude values with 0.0

preprocessing_dataset["Longitude"].fillna("0.0", inplace = True)  

In [26]:
# Verifying that all the null values in latitude column are replaced with 0.0

preprocessing_dataset.Latitude.isna().sum()

0

In [27]:
# Verifying that all the null values in longitude column are replaced with 0.0

preprocessing_dataset.Longitude.isna().sum()

0

### 1.2.3. Other variables

Not enough is known about the missing values in other variables. These empty entries are replaced with a question mark(?) for the time being.

In [28]:
# Calculating the number of missing values in the dataset

preprocessing_dataset.isna().sum()

Year              0
Mo              405
Dy              557
Hr             2031
Mn             2236
Sec            3353
Tsu            4291
Vol            6080
Addl              0
Name              0
Latitude          0
Longitude         0
Focal          2951
Mag            1783
MMI Int        3379
Num            4104
De             3634
Num.1          4932
De.1           4747
$Mill          5648
De.2           1757
Num.2          5375
De.3           4475
Num.3          5675
De.4           5250
Unnamed: 25    6083
dtype: int64

In [29]:
# Replacing the missing values with a question mark(?)

preprocessing_dataset = preprocessing_dataset.fillna('?')
print(preprocessing_dataset)

      Year Mo  Dy  Hr  Mn Sec  Tsu  Vol Addl  \
0    -2150  ?   ?   ?   ?   ?    ?    ?    *   
1    -2000  ?   ?   ?   ?   ?  Tsu    ?    *   
2    -2000  ?   ?   ?   ?   ?    ?    ?    *   
3    -1610  ?   ?   ?   ?   ?  Tsu  Vol    *   
4    -1566  ?   ?   ?   ?   ?    ?    ?    *   
...    ... ..  ..  ..  ..  ..  ...  ...  ...   
6139  2019  9  24  11   1  55    ?    ?    *   
6140  2019  9  25  23  46  44    ?    ?    *   
6141  2019  9  26  10  59  26    ?    ?    *   
6142  2019  9  26  16  36  18    ?    ?    *   
6143  2019  9  29  15  57  53    ?    ?    *   

                                  Name  ... De Num.1 De.1 $Mill De.2 Num.2  \
0         JORDAN: BAB-A-DARAA,AL-KARAK  ...  ?     ?    ?     ?    3     ?   
1                        SYRIA: UGARIT  ...  3     ?    ?     ?    ?     ?   
2                      TURKMENISTAN: W  ...  1     ?    ?     ?    1     ?   
3     GREECE: THERA ISLAND (SANTORINI)  ...  ?     ?    ?     ?    ?     ?   
4              ISRAEL: ARIHA (JER

In [30]:
# Verifying that there are no missing values present in the entire dataset

print(preprocessing_dataset.isna().sum())

Year           0
Mo             0
Dy             0
Hr             0
Mn             0
Sec            0
Tsu            0
Vol            0
Addl           0
Name           0
Latitude       0
Longitude      0
Focal          0
Mag            0
MMI Int        0
Num            0
De             0
Num.1          0
De.1           0
$Mill          0
De.2           0
Num.2          0
De.3           0
Num.3          0
De.4           0
Unnamed: 25    0
dtype: int64


## 1.3. Filtering required columns

Out of the 26 columns, Month, Day, Hour, Minute and Second that the earthquake occurred is thought to be not relevant for most of the analysis. They are referred to later when comparing the volcano and earthquake events. For the rest of the analysis, only the following columns are filtered: 

Year, Tsu, Vol, Name, Latitude, Longitude, Focal Depth, Mag, and MMI Int. The code for filtering these columns is displayed below:

In [31]:
dataset_for_analysis = preprocessing_dataset.filter(['Year','Tsu', 'Vol','Name','Latitude','Longitude','Focal Depth', 'Mag','MMI Int'], axis=1)
print(dataset_for_analysis)

      Year  Tsu  Vol                              Name Latitude Longitude  \
0    -2150    ?    ?      JORDAN: BAB-A-DARAA,AL-KARAK     31.1      35.5   
1    -2000  Tsu    ?                     SYRIA: UGARIT   35.683      35.8   
2    -2000    ?    ?                   TURKMENISTAN: W       38      58.2   
3    -1610  Tsu  Vol  GREECE: THERA ISLAND (SANTORINI)     36.4      25.4   
4    -1566    ?    ?           ISRAEL: ARIHA (JERICHO)     31.5      35.3   
...    ...  ...  ...                               ...      ...       ...   
6139  2019    ?    ?         PAKISTAN: MIRPUR DISTRICT   33.106    73.766   
6140  2019    ?    ?          INDONESIA: MALUKU: AMBON    -3.45   128.347   
6141  2019    ?    ?                  TURKEY: ISTANBUL    40.89    28.173   
6142  2019    ?    ?              CHILE: SOUTH CENTRAL  -40.815   -72.002   
6143  2019    ?    ?                 CHILE: CONCEPCION  -35.473   -73.162   

      Mag MMI Int  
0     7.3       ?  
1       ?      10  
2     7.1      

## 1.4. Adding required columns 

The 'Name' column consists of Country and Region separated by ":". Two new columns *Country* and *Region* are created by splitting the *Name* column into Country and Region for accurate analysis using the following code:

In [32]:
dataset_for_analysis['Country'] = dataset_for_analysis.Name.str.split(':').str[0]
dataset_for_analysis['Region'] = dataset_for_analysis.Name.str.split(':').str[1]
print(dataset_for_analysis)

      Year  Tsu  Vol                              Name Latitude Longitude  \
0    -2150    ?    ?      JORDAN: BAB-A-DARAA,AL-KARAK     31.1      35.5   
1    -2000  Tsu    ?                     SYRIA: UGARIT   35.683      35.8   
2    -2000    ?    ?                   TURKMENISTAN: W       38      58.2   
3    -1610  Tsu  Vol  GREECE: THERA ISLAND (SANTORINI)     36.4      25.4   
4    -1566    ?    ?           ISRAEL: ARIHA (JERICHO)     31.5      35.3   
...    ...  ...  ...                               ...      ...       ...   
6139  2019    ?    ?         PAKISTAN: MIRPUR DISTRICT   33.106    73.766   
6140  2019    ?    ?          INDONESIA: MALUKU: AMBON    -3.45   128.347   
6141  2019    ?    ?                  TURKEY: ISTANBUL    40.89    28.173   
6142  2019    ?    ?              CHILE: SOUTH CENTRAL  -40.815   -72.002   
6143  2019    ?    ?                 CHILE: CONCEPCION  -35.473   -73.162   

      Mag MMI Int       Country                     Region  
0     7.3     

## 1.5. Offsetting

The 'Year' column ranges from -2150 to 2019 as verified by the code below:

In [33]:
dataset_for_analysis['Year'].describe()

count    6144.000000
mean     1804.691243
std       376.404999
min     -2150.000000
25%      1820.000000
50%      1928.000000
75%      1988.000000
max      2019.000000
Name: Year, dtype: float64

The analysis of this dataset is aimed at splitting the dataset into 1000-year sets to observe trends within and between each set. In order to do this, 2150 is added to every row in *Year* to make the partitioning easier. This way, *Year* now ranges from 0 to 4160, making it easier to partition during the analysis. Offsetting is demonstrated in the code below:

In [34]:
UpdatedYear = dataset_for_analysis['Year'] + 2150
#print(UpdatedYear)
dataset_for_analysis['UpdatedYear'] = UpdatedYear
print(dataset_for_analysis)

      Year  Tsu  Vol                              Name Latitude Longitude  \
0    -2150    ?    ?      JORDAN: BAB-A-DARAA,AL-KARAK     31.1      35.5   
1    -2000  Tsu    ?                     SYRIA: UGARIT   35.683      35.8   
2    -2000    ?    ?                   TURKMENISTAN: W       38      58.2   
3    -1610  Tsu  Vol  GREECE: THERA ISLAND (SANTORINI)     36.4      25.4   
4    -1566    ?    ?           ISRAEL: ARIHA (JERICHO)     31.5      35.3   
...    ...  ...  ...                               ...      ...       ...   
6139  2019    ?    ?         PAKISTAN: MIRPUR DISTRICT   33.106    73.766   
6140  2019    ?    ?          INDONESIA: MALUKU: AMBON    -3.45   128.347   
6141  2019    ?    ?                  TURKEY: ISTANBUL    40.89    28.173   
6142  2019    ?    ?              CHILE: SOUTH CENTRAL  -40.815   -72.002   
6143  2019    ?    ?                 CHILE: CONCEPCION  -35.473   -73.162   

      Mag MMI Int       Country                     Region  UpdatedYear  
0

### Description of the columns

    Year     : The year in which earthquake occurred. It ranges from -2150 to 2019.
    Tsu      : Occurrence of tsunami
    Vol      : Occurrence of a volcano
    Name     : Place where the earthquake occurred
    Latitude : Latitude of earthquake occurrence
    Longitude: Longitude of the earthquake occurrence
    Mag      : Magntiude of the earthquake
    MMI Int  : Measure of damage caused by the earthquake

## 1.6. Downloading preprocessed dataset as a csv file

The following code downloads the preprocessed dataset as a csv file onto the local system. The downloaded file is then used for further analysis from here on.

In [35]:
dataset_for_analysis.to_csv('Dataset For Analysis.csv')

## 1.7. Insights

### Overview of the dataset

    1. The dataset consists of 26 columns and 6144 rows.

### Addressing missing values

    1. There exists one missing value in *Name* column. 
    2. The latitude and longitude did not match with any other record in the dataset, indicating a unique event.
    3. Google Maps was then used to identify the location as Cook Strait in New Zealand.
    4. The database was updated accordingly and verified to make sure the null value was updated.
    5. The missing values in latitude and longitude were replaced with 0.0 in accordance with NOAA's definitions. 
    6. The reasons behind the missing values in other variables is not well understood and is replaced with a '?'.
    
    
### Filtering and adding required columns

    1. Out of the entire dataset, eight columns: Year, Tsu, Vol, Name, Latitude, Longitude, Magnitude, MMI Int are filtered.
    2. 'Name' is of the format 'Country:Region'. It is split accordingly and two new columns are created for Country and Region
    3. The dataset now consists of 10 columns
    
### Offsetting

    1. The Year column ranges from -2150 to 2019. The negative sign indicates B.C. 
    2. In further analysis, the dataset will be partioned into subsets for detail exploration. 
    3. To do this, a new column UpdatedYear is creating by adding 2150 to every row in Year column.
    4. The UpdatedYear now ranges from 0 to 4169. 
    5. The dataset now consists of 11 columns. 